File: JavaScript/153-Find Minimum in Rotated Sorted Array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
    let left = 0, right = nums.length-1;
    while (left + 1 < right){
        mid = left + Math.floor((right-left)/2);
        if (nums[mid] <= nums[right]){
            right = mid;
        }else{
            left = mid;
        }
    }
    return Math.min(nums[left], nums[right]);
};

File: C++/153-Find Minimum in Rotated Sorted Array.cpp
class Solution {
public:
    int findMin(vector<int>& nums) {
        
        int N = nums.size();

        // when N == 1 or nums[N-1] > nums[0]
        if (nums[N-1] >= nums[0]) return nums[0];

        // find the first element < nums[0]
        int left = 0, right = N-1;
        while (left + 1 < right) {
            int mid = left + (right-left)/2;
            if (nums[mid] < nums[0]) {
                right = mid;
            } else {
                left = mid;
            }
        }
        
        if (nums[left] < nums[0]) {
            return nums[left];
        }
        return nums[right];
    }
};